pkg/webhook/cluster/mutation/mutation.go

```diff
@@ -0,0 +1,116 @@
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+// AdmissionHandler for mutating Kubermatic Cluster CRD.
+type AdmissionHandler struct {
+	log     logr.Logger
+	decoder *admission.Decoder
+}
+
+// NewAdmissionHandler returns a new cluster mutation AdmissionHandler.
+func NewAdmissionHandler() *AdmissionHandler {
+	return &AdmissionHandler{}
+}
+
+func (h *AdmissionHandler) InjectLogger(l logr.Logger) error {
+	h.log = l.WithName("cluster-mutation-handler")
+	return nil
+}
+
+func (h *AdmissionHandler) InjectDecoder(d *admission.Decoder) error {
+	h.decoder = d
+	return nil
+}
+
+func (h *AdmissionHandler) Handle(ctx context.Context, req webhook.AdmissionRequest) webhook.AdmissionResponse {
+	cluster := &kubermaticv1.Cluster{}
+	oldCluster := &kubermaticv1.Cluster{}
+
+	switch req.Operation {
+	case admissionv1.Create:
+		// NOP
+	case admissionv1.Update:
+		if err := h.decoder.Decode(req, cluster); err != nil {
+			return admission.Errored(http.StatusBadRequest, err)
+		}
+		if err := h.decoder.DecodeRaw(req.OldObject, oldCluster); err != nil {
+			return admission.Errored(http.StatusBadRequest, err)
+		}
+
+		if err := h.mutateUpdate(ctx, oldCluster, cluster); err != nil {
+			h.log.Info("cluster mutation failed", "error", err)
+			return webhook.Errored(http.StatusInternalServerError, fmt.Errorf("cluster mutation request %s failed: %v", req.UID, err))
+		}
+
+		mutatedCluster, err := json.Marshal(cluster)
+		if err != nil {
+			return webhook.Errored(http.StatusInternalServerError, fmt.Errorf("marshaling cluster object failed: %v", err))
+		}
+
+		return admission.PatchResponseFromRaw(req.Object.Raw, mutatedCluster)
+	case admissionv1.Delete:
+		// NOP
+	default:
+		return admission.Errored(http.StatusBadRequest, fmt.Errorf("%s not supported on cluster resources", req.Operation))
+	}
+	return webhook.Allowed(fmt.Sprintf("no mutation done for request %s", req.UID))
+}
+
+func (h *AdmissionHandler) mutateUpdate(ctx context.Context, oldCluster, newCluster *kubermaticv1.Cluster) error {
+	// This part of the code handles the CCM/CSI migration. It currently works
+	// only for OpenStack clusters, in the following way:
+	//   * Add the CCM/CSI migration annotations
+	//   * Enable the UseOctaiva flag
+	switch {
+	case newCluster.Spec.Cloud.Openstack != nil:
+		if v, ok := newCluster.Spec.Features[kubermaticv1.ClusterFeatureExternalCloudProvider]; ok && v {
+			if oldV, oldOk := oldCluster.Spec.Features[kubermaticv1.ClusterFeatureExternalCloudProvider]; !oldOk || !oldV {
+				useOctavia := true
+				if newCluster.ObjectMeta.Annotations == nil {
+					newCluster.ObjectMeta.Annotations = map[string]string{}
+				}
+
+				newCluster.ObjectMeta.Annotations[kubermaticv1.CCMMigrationNeededAnnotation] = ""
+				newCluster.ObjectMeta.Annotations[kubermaticv1.CSIMigrationNeededAnnotation] = ""
+				newCluster.Spec.Cloud.Openstack.UseOctavia = &useOctavia
```

Comment:
nit: you could use
https://github.com/kubernetes/utils/blob/fddb29f9d0095a7a37582a1d86732ae83620f2e8/pointer/pointer.go#L77-L80
to avoid declaring `useOctavia` variable